### PR TITLE
Optimize {run_add_to,run_remove_from}_scope_queue

### PR DESCRIPTION
**Pull request by alexkirsz**
This PR does two things:
* avoids allocating a new vector in `add_to_scope_internal_shallow`/`remove_from_scope_internal_shallow`
* use an actual queue instead of a stack for these operations. This part doesn't change much, if anything.

This shows improvements up to ~6% on a 5000 modules SSR benchmark:
image

On 20k modules, there are no significant changes.

### CODE REVIEW

File: crates/turbo-tasks-memory/src/task.rs [hunk 1]
```diff
@@ -657,66 +657,70 @@ impl Task {
...
                         }
                     }
-                    let children = children.iter().copied().collect::<Vec<_>>();
+                }
 
-                    // add to dirty list of the scope (potentially schedule)
-                    let schedule_self =
-                        self.add_self_to_new_scope(&mut state, id, backend, turbo_tasks);
-                    drop(state);
+                queue.extend(children.iter().copied().map(|child| (child, depth + 1)));
```
Comment (alexkirsz) [inline crates/turbo-tasks-memory/src/task.rs:712]: This line is the actual change.

File: crates/turbo-tasks-memory/src/task.rs [hunk 2]
```diff
@@ -1460,81 +1443,54 @@ impl Task {
...
+                id,
+                is_optimization_scope,
+                depth,
+                backend,
+                turbo_tasks,
+                &mut queue,
+            );
+        });
+        if queue.len() > SPLIT_OFF_QUEUE_AT {
+            let split_off_queue = queue.split_off(SPLIT_OFF_QUEUE_AT);
```
Comment (alexkirsz) [inline crates/turbo-tasks-memory/src/task.rs:1470]: This logic is now much simpler.

File: crates/turbo-tasks-memory/src/task.rs [hunk 3]
```diff
@@ -859,17 +860,11 @@ impl Task {
             TaskScopes::Inner(ref mut set, _) => {
                 if set.remove(id) {
                     self.remove_self_from_scope(&mut state, id, backend, turbo_tasks);
-                    let children = state.children.iter().copied().collect::<Vec<_>>();
+                    queue.extend(state.children.iter().copied());
```
Comment (alexkirsz) [inline crates/turbo-tasks-memory/src/task.rs:863]: This line is the actual change.

**Review by jridgewell — APPROVED** [review]
Much better!

**Review by ForsakenHarmony — APPROVED** [review]

**Pull request merged**